go/metaparticle/docker_impl.go

```diff
@@ -1,81 +1,288 @@
+		return errors.Wrap(err, "Error reading push output from docker")
 	}
-	cmd := exec.Command("docker", "push", image)
-	cmd.Stdout = stdout
-	cmd.Stderr = stderr
 
-	return cmd.Run()
+	return nil
 }
 
+func parsePorts(ports []int32) (nat.PortMap, nat.PortSet, error) {
+	portBindings := make(nat.PortMap)
+	exposedPorts := make(nat.PortSet)
+	for _, port := range ports {
+		if port <= 0 || port > 65535 {
+			return nil, nil, errInvalidPort(port)
+		}
+		// TODO: some people might need UDP ports, will probably need to switch the slice of int32 to something else
+		natPort, err := nat.NewPort("tcp", strconv.Itoa(int(port)))
+		exposedPorts[natPort] = struct{}{}
+		if err != nil {
+			return nil, nil, errors.Wrap(err, "Error converting the port number")
+		}
+		portBinding := []nat.PortBinding{{HostPort: strconv.Itoa(int(port))}}
+
+		portBindings[natPort] = portBinding
+	}
+
+	return portBindings, exposedPorts, nil
+}
+
+// Run creates and starts a container with the given image and name, and runtime options (e.g. exposed ports) specified in the config parameter
 func (d *DockerImpl) Run(image string, name string, config *Runtime, stdout io.Writer, stderr io.Writer) error {
 	if len(image) == 0 {
-		return fmt.Errorf("An image must be specified")
+		return errEmptyImageName
 	}
 
 	if len(name) == 0 {
-		return fmt.Errorf("The container's name must be specified")
+		return errEmptyContainerName
 	}
 
 	if config == nil {
-		return fmt.Errorf("config must not be nil")
+		return errNilRuntimeConfig
 	}
-	cmdName := "docker"
-	cmdParams := []string{"run", "-d"}
-	for _, port := range config.Ports {
-		cmdParams = append(cmdParams, "-p", fmt.Sprintf("%d:%d", port, port))
+
+	portBindings, exposedPorts, err := parsePorts(config.Ports)
+	if err != nil {
+		return errors.Wrap(err, "Error parsing configuration ports")
 	}
-	cmdParams = append(cmdParams, "--name", name, image)
 
-	cmd := exec.Command(cmdName, cmdParams...)
-	cmd.Stdout = stdout
-	cmd.Stderr = stderr
+	ctx := context.Background()
+	if _, err := d.containerRunner.ContainerCreate(ctx, &container.Config{Image: image, ExposedPorts: exposedPorts},
+		&container.HostConfig{PortBindings: portBindings}, nil, name); err != nil {
+		return errors.Wrap(err, "Error creating the container")
+	}
 
-	return cmd.Run()
+	if err := d.containerRunner.ContainerStart(ctx, name, types.ContainerStartOptions{}); err != nil {
+		return errors.Wrap(err, "Error starting the container")
+	}
+	return nil
 }
 
+// Logs attaches to the container with the given name and prints the log to stdout
 func (d *DockerImpl) Logs(name string, stdout io.Writer, stderr io.Writer) error {
 	if len(name) == 0 {
-		return fmt.Errorf("A container name must be specified")
+		return errEmptyContainerName
+	}
+	ctx := context.Background()
+	res, err := d.containerRunner.ContainerLogs(ctx, name, types.ContainerLogsOptions{ShowStdout: true, ShowStderr: true, Follow: true})
+	defer res.Close()
+	if err != nil {
```

Comment:
probably want to check this error before the `defer` call above? if not `res` might be `nil`?